File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Hardware/Motors/Encoder.java
package org.firstinspires.ftc.teamcode.Hardware.Motors;

import com.qualcomm.robotcore.hardware.DcMotorEx;

public class Encoder {
    private DcMotorEx encoder;
    private int pos = 0;
    public Encoder(DcMotorEx encoder) {
        this.encoder = encoder;
    }


    public void encoder(DcMotorEx encoder){
        this.encoder = encoder;
        return;
    }
    public void updateEncoder(){
        pos = encoder.getCurrentPosition();
    }


    public void resetEncoder(){
        pos = 0;
    }
    public int getCurrentPosition(){
        updateEncoder();
        return pos;
    }
    public double getVelocity(){
        return encoder.getVelocity();
    }



}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/Encoder.kt
package org.firstinspires.ftc.teamcode.hardware

import com.qualcomm.robotcore.hardware.DcMotor

interface Encoder {
    var isFlipped: Boolean
    val flippedFactor: Int
        get() = if (isFlipped) -1 else 1

    fun getCurrentPosition(): Int
    fun reset()
}

class MotorEncoder(val motor: DcMotor): Encoder {

    private var isFlippedBacking = false
    override var isFlipped: Boolean
        get() = isFlippedBacking
        set(value) { isFlippedBacking = value }

    override fun getCurrentPosition(): Int = motor.currentPosition * flippedFactor
    override fun reset() {
        val modeBefore = motor.mode
        motor.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        motor.mode = modeBefore
    }
}